core/templates/components/state-directives/outcome-editor/outcome-editor.component.spec.ts

```diff
@@ -0,0 +1,398 @@
+      $scope: $scope
+    }, {
+      isEditable: () => true,
+      getOnSaveCorrectnessLabelFn: () => {
+        return () => {};
+      },
+      showMarkAllAudioAsNeedingUpdateModalIfRequired: () => {},
+      getOnSaveFeedbackFn: () => {
+        return () => {};
+      },
+      getOnSaveDestFn: () => {
+        return () => {};
+      }
+    });
+
+    spyOn(StateEditorService, 'isExplorationWhitelisted').and.returnValue(true);
+  }));
+
+  afterEach(() => {
+    ctrl.$onDestroy();
+  });
+
+  it('should set component properties on initialization', () => {
+    let outcome = {
+      feedback: {
+        html: '<p> Previous HTML string </p>'
+      },
+      hasNonemptyFeedback: () => true
+    };
+    ctrl.outcome = outcome;
+
+    expect(ctrl.editOutcomeForm).toEqual(undefined);
+    expect(ctrl.canAddPrerequisiteSkill).toBe(undefined);
+    expect(ctrl.feedbackEditorIsOpen).toBe(undefined);
+    expect(ctrl.destinationEditorIsOpen).toBe(undefined);
+    expect(ctrl.correctnessLabelEditorIsOpen).toBe(undefined);
+    expect(ctrl.savedOutcome).toBe(undefined);
+
+    ctrl.$onInit();
+
+    expect(ctrl.editOutcomeForm).toEqual({});
+    expect(ctrl.canAddPrerequisiteSkill).toBe(false);
+    expect(ctrl.feedbackEditorIsOpen).toBe(false);
+    expect(ctrl.destinationEditorIsOpen).toBe(false);
+    expect(ctrl.correctnessLabelEditorIsOpen).toBe(false);
+    expect(ctrl.savedOutcome).toEqual(outcome);
+  });
+
+  it('should save feedback on external save event when editFeedbackForm is' +
+    ' valid and state is not invalid after feedback save', () => {
+    let onExternalSaveEmitter = new EventEmitter();
+    spyOnProperty(ExternalSaveService, 'onExternalSave')
+      .and.returnValue(onExternalSaveEmitter);
+    spyOn(ctrl, 'invalidStateAfterFeedbackSave').and.returnValue(false);
+    spyOn(ctrl, 'saveThisFeedback');
+
+    ctrl.$onInit();
+
+    ctrl.feedbackEditorIsOpen = true;
+    ctrl.editOutcomeForm = {
+      editFeedbackForm: {
+        $valid: true
+      }
+    };
+
+    onExternalSaveEmitter.emit();
+    $scope.$apply();
+
+    expect(ctrl.saveThisFeedback).toHaveBeenCalledWith(false);
+  });
+
+  it('should cancel feedback edit on external save event when' +
+    ' editFeedbackForm is not valid or state us not valid after' +
+    ' feedback save', () => {
+    let onExternalSaveEmitter = new EventEmitter();
+    spyOnProperty(ExternalSaveService, 'onExternalSave')
+      .and.returnValue(onExternalSaveEmitter);
+    spyOn(ctrl, 'invalidStateAfterFeedbackSave').and.returnValue(false);
+    spyOn(ctrl, 'cancelThisFeedbackEdit');
+
+    ctrl.$onInit();
+
+    ctrl.feedbackEditorIsOpen = true;
+    ctrl.editOutcomeForm = {
+      editFeedbackForm: {
+        $valid: false
+      }
+    };
+
+    onExternalSaveEmitter.emit();
+    $scope.$apply();
+
+    expect(ctrl.cancelThisFeedbackEdit).toHaveBeenCalled();
+  });
+
+  it('should save destination on interaction change when edit destination' +
```

Comment:
Probably put a comment mentioning the steps (setup, pre-check, action, post-check)

Reply:
Added comments